frame/paged-list/src/paged_list.rs

```diff
@@ -0,0 +1,530 @@
+	pub fn delete<Prefix: StorageInstance, Hasher: StorageHasher>(&self) {
+		delete_page::<Prefix, Hasher>(self.index);
+	}
+}
+
+/// Delete a page with `index` from storage.
+// Does not live under `Page` since it does not require the `Value` generic.
+pub(crate) fn delete_page<Prefix: StorageInstance, Hasher: StorageHasher>(index: PageIndex) {
+	let key = page_key::<Prefix, Hasher>(index);
+	sp_io::storage::clear(key.as_ref());
+}
+
+/// Storage key of a page with `index`.
+// Does not live under `Page` since it does not require the `Value` generic.
+pub(crate) fn page_key<Prefix: StorageInstance, Hasher: StorageHasher>(
+	index: PageIndex,
+) -> Hasher::Output {
+	(StoragePagedListPrefix::<Prefix>::final_prefix(), b"page", index).using_encoded(Hasher::hash)
+}
+
+impl<V: Clone> Iterator for Page<V> {
+	type Item = V;
+
+	fn next(&mut self) -> Option<Self::Item> {
+		let val = self.values.get(self.next as usize).cloned();
+		self.next.saturating_inc();
+		val
+	}
+}
+
+/// Iterates over values of a [`StoragePagedList`].
+///
+/// Can optionally drain the iterated values.
+pub struct StoragePagedListIterator<Prefix, Hasher, Value, ValuesPerPage> {
+	// Design: we put the Page into the iterator to have fewer storage look-ups. Yes, these
+	// look-ups would be cached anyway, but bugging the overlay on each `.next` call still seems
+	// like a poor trade-off than caching it in the iterator directly. Iterating and modifying is
+	// not allowed at the same time anyway, just like with maps. Note: if Page is empty then
+	// the iterator did not find any data upon setup or ran out of pages.
+	page: Option<Page<Value>>,
+	drain: bool,
+	meta: StoragePagedListMeta<Prefix, Hasher, Value, ValuesPerPage>,
+}
+
+impl<Prefix, Hasher, Value, ValuesPerPage>
+	StoragePagedListIterator<Prefix, Hasher, Value, ValuesPerPage>
+where
+	Prefix: StorageInstance,
+	Value: FullCodec + Clone,
+	Hasher: StorageHasher,
+	ValuesPerPage: Get<u32>,
+{
+	/// Read self from the storage.
+	pub fn from_meta(
+		meta: StoragePagedListMeta<Prefix, Hasher, Value, ValuesPerPage>,
+		drain: bool,
+	) -> Self {
+		let mut page = Page::<Value>::from_storage::<Prefix, Hasher>(meta.first_page);
+
+		if let Some(ref mut page) = page {
+			page.next = meta.first_value;
+		}
+
+		Self { page, drain, meta }
+	}
+}
+
+impl<Prefix, Hasher, Value, ValuesPerPage> Iterator
+	for StoragePagedListIterator<Prefix, Hasher, Value, ValuesPerPage>
+where
+	Prefix: StorageInstance,
+	Value: FullCodec + Clone,
+	Hasher: StorageHasher,
+	ValuesPerPage: Get<u32>,
+{
+	type Item = Value;
+
+	fn next(&mut self) -> Option<Self::Item> {
+		let page = self.page.as_mut()?;
+
+		if let Some(val) = page.next() {
+			if self.drain {
+				self.meta.first_value.saturating_inc();
+				self.meta.store()
+			}
+			return Some(val)
+		}
+		if self.drain {
+			// page is empty
+			page.delete::<Prefix, Hasher>();
+			self.meta.first_value = 0;
+			self.meta.first_page.saturating_inc();
+			self.meta.store();
+		}
```

Comment:
Wouldn't it be better to immediately delete the page after its last value is fetched, instead of deferring it to the next time `next` is called? For example, if someone doesn't call `next` until it returns `None` but just instantiates the iterator, calls `next()` once, and then drops the iterator, and the value that was fetched was the last one in a given page, then the `first_value` in storage will point to a value which doesn't exist.

Reply:
Yes i will try again to refactor this stuff into smaller functions. Then it should directly delete the page when it becomes empty.

Reply:
Ah nice, this made the code simpler and easier to rule about, thanks!